src/contracts/core/AllocationManager.sol

```diff
@@ -0,0 +1,731 @@
+     * @param numToComplete the number of pending free magnitudes deallocations to complete
+     * @dev read through pending free magnitudes and add to freeMagnitude if completableTimestamp is >= block timestamp
+     * In addition to updating freeMagnitude, updates next starting index to read from for pending free magnitudes after completing
+     */
+    function _updateFreeMagnitude(address operator, IStrategy strategy, uint16 numToComplete) internal {
+        OperatorMagnitudeInfo memory info = operatorMagnitudeInfo[operator][strategy];
+        
+        (uint64 freeMagnitudeToAdd, uint192 completed) =
+            _getPendingFreeMagnitude(operator, strategy, numToComplete, info.nextPendingFreeMagnitudeIndex);
+        info.freeMagnitude += freeMagnitudeToAdd;
+        info.nextPendingFreeMagnitudeIndex += completed;
+
+        operatorMagnitudeInfo[operator][strategy] = info;
+    }
+
+    /**
+     * @notice For a single strategy, modify magnitude allocations for each of the specified operatorSets
+     * @param operator address to modify allocations for
+     * @param allocation the magnitude allocations to modify for a single strategy
+     * @param allocationEffectTimestamp the timestamp when the allocations will take effect
+     * @param deallocationCompletableTimestamp the timestamp when the deallocations will be completable
+     * @dev For each allocation, allocation.operatorSets MUST be ordered in ascending order according to the
+     * encoding of the operatorSet. This is to prevent duplicate operatorSets being passed in. The easiest way to ensure
+     * ordering is to sort allocated operatorSets by address first, and then sort for each avs by ascending operatorSetIds.
+     */
+    function _modifyAllocations(
+        address operator,
+        MagnitudeAllocation calldata allocation,
+        uint32 allocationEffectTimestamp,
+        uint32 deallocationCompletableTimestamp
+    ) internal {
+        require(allocation.operatorSets.length == allocation.magnitudes.length, InputArrayLengthMismatch());
+
+        for (uint256 i = 0; i < allocation.operatorSets.length; ++i) {
+            require(
+                avsDirectory.isOperatorSet(allocation.operatorSets[i].avs, allocation.operatorSets[i].operatorSetId),
+                InvalidOperatorSet()
+            );
+
+            bytes32 operatorSetKey = _encodeOperatorSet(allocation.operatorSets[i]);
+            // prep magnitudeUpdates in storage
+
+            // Check that there is at MOST `MAX_PENDING_UPDATES` combined allocations & deallocations for the operator, operatorSet, strategy
+            (uint256 numPendingAllocations, uint256 numPendingDeallocations) =
+                _getPendingAllocationsAndDeallocations(operator, allocation.strategy, operatorSetKey);
+
+            require(
+                numPendingAllocations + numPendingDeallocations < MAX_PENDING_UPDATES,
+                PendingAllocationOrDeallocation()
+            );
```

Comment:
We only allow a single pending update, so this is extremely complex for no upside.

Instead:
- Keep confirmed allocations/deallocations in a `History[]` type structure, since we need historical lookup in some cases
- Keep _the current pending alloc/dealloc_ in a separate state variable and push it to the `History[]` when it's confirmed

... then this can be changed to just read the pending variable rather than needing to do an array-based lookup.

Reply:
Also this `require` statement and the comment that says we're checking that there are "at most `MAX_PENDING_UPDATES`" is just incorrect.

The only point of these two lines is, roughly:

`require(noPendingAllocOrDealloc)`

... this is very unclear from the code. We need to rewrite this to reflect the fact that we are not supporting multiple alloc/dealloc anymore -- the legacy handling for multiple magnitude changes is extremely unnecessary.

Reply:
Fixed with Yash's refactor